### PR TITLE
Improve the performance of query string encoding/decoding

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

As demonstrated by @franz1981, we can bypass JVM's array boundary checks
completely, by using `PlatformDependent.getByte()`.

Modifications:

- Optimize `QueryStringEncoder.isUnsafeOctet()` and
  `QueryStringDecoder.decodeHexNibble()` using
  `PlatformDependent.getByte()`.
- Fix an off-by-one error in `decodeHexNibble()`

Result:

- Another performance gain
- Fixes an off-by-one error in `decodeHexNibble()`

### CODE REVIEW

**Comment by trustin** [comment]
This PR also fixes the off-by-one error :scream:

File: core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java [hunk 1]
```diff
@@ -198,15 +199,17 @@ private static int indexOfUnsafeOctet(String s, int start) {
...
 
-    // Do not inline `index` or remove `index < 0` check.
-    // They are specifically written to tell JVM to optimize away the array boundary check.
-    @SuppressWarnings({ "ConstantConditions", "UnnecessaryLocalVariable" })
     private static boolean isUnsafeOctet(char c) {
-        final int index = c;
-        if (index < 0 || index >= SAFE_OCTETS.length) {
-            return true;
+        if (c >= SAFE_OCTETS.length) {
+            return false;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java]: The original logic is a bug?
It returns `true` if `index >= SAFE_OCTETS.length`. 🤔

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java]: > This PR also fixes the off-by-one error 😱

Is this? 🤣

File: core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java [hunk 2]
```diff
@@ -201,7 +201,7 @@ private static int indexOfUnsafeOctet(String s, int start) {
 
     private static boolean isUnsafeOctet(char c) {
         if (c >= SAFE_OCTETS.length) {
-            return false;
+            return true;
```
Comment (franz1981) [inline core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java]: sorry, missed that :O

File: core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java (hunk 1 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java]: Nope. We had to return `true` here :scream:

File: core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java (hunk 2 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java]: :rofl:

File: core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java [hunk 3]
```diff
@@ -264,15 +266,17 @@ private static int decodeHexByte(char c1, char c2) {
         return (hi << 4) | lo;
     }
 
-    // Do not inline `index` or remove `index < 0` check.
-    // They are specifically written to tell JVM to optimize away the array boundary check.
-    @SuppressWarnings({ "ConstantConditions", "UnnecessaryLocalVariable" })
     private static int decodeHexNibble(char c) {
-        final int index = c;
-        if (index < 0 || index > OCTETS_TO_HEX.length) {
+        if (c >= OCTETS_TO_HEX.length) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java]: Off-by-one error here

**Comment by trustin** [comment]
Explicit widening from `char` to `int` seems unnecessary.

**Comment by trustin** [comment]
Quick benchmarking tells me explicit widening helps a little bit actually. Let me update.

**Comment by ikhoon** [comment]
Still LGTM

**Pull request merged**